[static runtime] Split out graph preparation from runtime

bwasti opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #44130 [static runtime] Add NNC
* **#44131 [static runtime] Split out graph preparation from runtime**

Differential Revision: [D23604305](https://our.internmc.facebook.com/intern/diff/D23604305)

bwasti committed:
[static runtime] Split out graph preparation from runtime

[ghstack-poisoned]

bwasti committed:
Update on "[static runtime] Split out graph preparation from runtime"




[ghstack-poisoned]

bwasti committed:
Update on "[static runtime] Split out graph preparation from runtime"




[ghstack-poisoned]

bwasti committed:
Update on "[static runtime] Split out graph preparation from runtime"




[ghstack-poisoned]

yinghai commented on benchmarks/static_runtime/deep_wide_pt_bench.cc:
```diff
@@ -75,6 +76,34 @@ static void BM_deep_wide_static(benchmark::State& state) {
+    runtime.run(inputs);
+  }
+
+  if (state.thread_index == 0) {
```
What's this?

bwasti committed:
Update on "[static runtime] Split out graph preparation from runtime"




[ghstack-poisoned]

hlu1 commented on torch/csrc/jit/runtime/static/impl.cpp:
```diff
@@ -46,14 +42,27 @@ StaticRuntime::StaticRuntime(const torch::jit::Module& m)
-      graph_->eraseInput(0);
+  if (g->inputs().at(0)->type()->is_module()) {
+    if (!g->inputs().at(0)->hasUses()) {
+      g->eraseInput(0);
```
Here Input(0) is only removed in the case that it does not have any use. Do you know cases that Input(0) does have uses later in the graph?

dzhulgakov commented on torch/csrc/jit/runtime/static/impl.cpp:
```diff
@@ -66,19 +75,13 @@ StaticRuntime::StaticRuntime(const torch::jit::Module& m)
-    CHECK((graph_->inputs().at(0)->type()->is_module()));
-    workspace_[graph_->inputs()[0]] = module_._ivalue();
-  }
+  TORCH_INTERNAL_ASSERT(graph_->inputs().size() == inps.size());
```
TORCH_CHECK is user-introduced error and INTERNAL_ASSERT something that shouldn't happen. The latter says "report bug to pytorch" in the message :)

hlu1 replied:
Got it.

dzhulgakov commented on torch/csrc/jit/runtime/static/impl.h:
```diff
@@ -32,12 +33,11 @@ class TORCH_API StaticRuntime {
   // Static runtime states
   // Value table (including weights)
-  ConstantMap workspace_;
+  mutable ConstantMap workspace_;
```
hm... is it because you use operator[]? maybe use at() then? and as I mentioned elsewhere, using a map is pretty inefficient, we should just remap everything into indices.

hlu1 replied:
Is there any documentation on how much this affect the overall latency? Replacing it with mapping to indices would certainly be net positive. I'm just wondering in terms of framework overhead, what kind of percentage this contribute.

bwasti committed:
Update on "[static runtime] Split out graph preparation from runtime"


Differential Revision: [D23604305](https://our.internmc.facebook.com/intern/diff/D23604305)

[ghstack-poisoned]
